File: src/templates/scripts/action_client.py
#!/usr/bin/env python3
import rospy
import actionlib
from custom_msgs.msg import ExampleAction, ExampleGoal

class TemplateActionClient:

    def __init__(self) -> None:
        goal = rospy.get_param(param_name='goal')

        self.action_client = actionlib.ActionClient(ns='/example_action', ActionSpec=ExampleAction)

        self.goal_handles = {}
        self.goal_handles['1'] = self.send_goal(goal=goal)

    def send_goal(self, goal):
        self.action_client.wait_for_server()

        goal_msg = ExampleGoal()
        goal_msg.goal = goal

        goal_handle = self.action_client.send_goal(
            goal=goal_msg,
            transition_cb=self.transition_callback,
            feedback_cb=self.feedback_callback
        )

        return goal_handle

    def transition_callback(self, goal_handle):
        # https://docs.ros.org/en/api/actionlib/html/classactionlib_1_1action__client_1_1CommState.html
        
        goal_id = {
            0 : 'PENDING',
            1 : 'ACTIVE',
            2 : 'CANCELED',
            3 : 'SUCCEEDED',
            4 : 'ABORTED',
            5 : 'REJECTED',
            6 : 'CANCELING',
            7 : 'RECALLING',
            8 : 'RECALLED',
            9 : 'LOST'
        }
        
        for i in self.goal_handles:

            if self.goal_handles[i] == goal_handle:
                break

        if goal_handle.get_comm_state() == 2:
            rospy.loginfo('Goal is active')

        if goal_handle.get_comm_state() == 7:
            rospy.loginfo('Goal is done')

            result = goal_handle.get_result()
            status = goal_handle.get_terminal_state()

            if status == 3:
                rospy.loginfo(msg=f'Goal succeeded: {result.result}')
            else:
                rospy.logerr(msg=f'Goal failed with status: {goal_id[status]}')

            rospy.signal_shutdown(reason='')

    def feedback_callback(self, goal_handle, feedback):
        rospy.loginfo(msg=f'Feedback: {feedback.feedback}')

if __name__ == '__main__':
    rospy.init_node(name='action_client')
    TemplateActionClient()
    rospy.spin()

File: src/templates/src/action_client.cpp
#include <ros/ros.h>
#include <actionlib/client/action_client.h>
#include <custom_msgs/ExampleAction.h>

typedef actionlib::ClientGoalHandle<custom_msgs::ExampleAction> GoalHandle;

class TemplateActionClient
{
    public:

    std::map<int, GoalHandle*> goal_handles;

    TemplateActionClient() : action_client("/example_action")
    {
        action_client.waitForActionServerToStart();
    }

    void transition_callback(const GoalHandle goal_handle)
    {        
        int index = 0;
        std::map<int, GoalHandle*>::iterator i = goal_handles.begin();
        while (i != goal_handles.end())
        {
            if (*(i->second) == goal_handle)
            {
                index = i->first;
                break;
            }
            i++;
        }

        actionlib::CommState CommState = goal_handle.getCommState();
        
        if (CommState.toString() == "ACTIVE")
        {
            ROS_INFO("Goal is active");
        }

        if (CommState.toString() == "DONE")
        {
            ROS_INFO("Goal is done");

            custom_msgs::ExampleResultConstPtr result = goal_handle.getResult();
            actionlib::TerminalState status = goal_handle.getTerminalState();

            if (status.toString() == "SUCCEEDED")
            {
                std::string res = (result->result == goal_msg.goal) ? "true" : "false";
                ROS_INFO("Goal succeeded: %s", res.c_str());
            }
            else
            {
                ROS_ERROR("Goal failed with status: %s", status.toString().c_str());
            }

            ros::shutdown();
        }
    }

    void feedback_callback
    (
        const GoalHandle goal_handle,
        const custom_msgs::ExampleFeedbackConstPtr &feedback
    )
    {
        ROS_INFO("Feedback: %i", (int)feedback->feedback);
    }

    GoalHandle send_goal(int goal)
    {
        goal_msg.goal = goal;
        GoalHandle goal_handle = action_client.sendGoal
        (
            goal_msg,
            boost::bind(&TemplateActionClient::transition_callback, this, _1),
            boost::bind(&TemplateActionClient::feedback_callback, this, _1, _2)
        );
        
        return goal_handle;
    }

    private:

    actionlib::ActionClient<custom_msgs::ExampleAction> action_client;
    custom_msgs::ExampleGoal goal_msg;
};

int main(int argc, char **argv)
{
    ros::init(argc, argv, "action_client");
    
    ros::AsyncSpinner spinner(5);
    spinner.start();

    TemplateActionClient action_client;

    int goal;
    ros::param::get("goal", goal);

    GoalHandle goal_handle = action_client.send_goal(goal);
    action_client.goal_handles[1] = &goal_handle;
    
    ros::waitForShutdown();

    return 0;
}